### ISSUE TITLE
Reserved document tech task causing other pawns to idle

### ISSUE THREAD

**Issue by archielecroy** [jptrrs/HumanResources#107]
Pawns assigned to document a knowledge and with study as the highest assigned priority are idling when a bill is reserved by, but not actively being worked on by another pawn. Idle pawns do not attempt to do other available but lower-priority tasks. Idling pawns still do unassignable, high-priority tasks. e.g.: wander, eat, sleep, recreation, and opportunistically gather.  
In my situation Red has started documenting a knowledge, then gone to higher priority tasks. Zy and Light are idling while tasks are available.

No messages in debug log.
Using Dubs performance analyzer, I see a JobGiver_Work entry on the idle pawn fire whenever the pawn is undrafted or reaches its wandering destination: "DocumentTech - HR_Learn - [JPT] Human Resources - Zy

My library setup: 1 simple research bench and 2 study desks, all linked to 3 book shelves. 
Free space remaining in library: 66  

First study desk has 2 bills:
1. Study tech: 
&nbsp;&nbsp;&nbsp;&nbsp;do forever
&nbsp;&nbsp;&nbsp;&nbsp;any worker
2. Document tech: 
&nbsp;&nbsp;&nbsp;&nbsp;do forever
&nbsp;&nbsp;&nbsp;&nbsp;any worker
&nbsp;&nbsp;&nbsp;&nbsp;reserved worker: Red  
  
  
Second study desk has 2 bills:
1. Study tech: 
&nbsp;&nbsp;&nbsp;&nbsp;do forever
&nbsp;&nbsp;&nbsp;&nbsp;any worker
2. Document tech: 
&nbsp;&nbsp;&nbsp;&nbsp;do forever
&nbsp;&nbsp;&nbsp;&nbsp;any worker
&nbsp;&nbsp;&nbsp;&nbsp; 
  
  
Actions that allow idle colonists to begin other tasks:  
1. Lowering idle colonists' study priority. 
&nbsp;&nbsp;&nbsp;&nbsp;Allows tasks that are now higher priority than study
2. Removing idle colonists' study priority. 
&nbsp;&nbsp;&nbsp;&nbsp;Allows all assigned tasks
3. Forcing the reserving pawn to actively work on the reserved bill
&nbsp;&nbsp;&nbsp;&nbsp;Allows all assigned tasks (including study tasks)
4. Removing idle pawns' document knowledge orders
&nbsp;&nbsp;&nbsp;&nbsp;Allows all assigned tasks (including study technology)
5. Editing the reserved bill to allow only the reserving pawn
&nbsp;&nbsp;&nbsp;&nbsp;Allows all assigned tasks (including study tasks)  
  
  

Actions that did not allow idle colonists to begin other tasks:  
1. (Un)suspending the unreserved Document Technology bill on the unreserved study bench
2. Deleting the unreserved Document Technology bill on the unreserved study bench
3. Creating another Document Technology bill on either study bench
4. Destroying the unreserved study bench

[Modlist](https://gist.github.com/archielecroy/cb03663a9a8a7921b8ecbba491321941)

**Comment by jptrrs** [comment]
Please, while the issue is going on, press Ctrl+F12 and send the game log. Thanks.

**Comment by archielecroy** [comment]
[game log 1](https://gist.github.com/5b20aa96c06bab5fdd9a1c5cb126ffe5)
In this log, pawns Red and Gabriela are idling, even though they're both assigned to Construct and there are active deconstruct orders. I repeatedly drafted and undrafted them to try and have them start a task, but they only take recreation tasks.

**Comment by archielecroy** [comment]
[game log 2](https://gist.github.com/archielecroy/793607a6e2e4e2f15e4e52ddcc5fe730)
Reproduced with a new save with minimal mods. 

Tribal start
Used god mode to create 2 study desk, a research bench, and a book shelf.
On each study desk added a study tech bill and then a document tech bill.
suspended one of the document tech bills.
Gave each pawn a document knowledge order via their Tech tab
Assigned each pawn to Study priority 1 via the Work tab
Pawn Gabela started documenting Devilstrand.
Paused, then drafted Gabela.
Removed Gabela's Study assignment via the Work tab.
Assigned all pawns to construction at low priority.
Spawned stacks of wood and ordered wood walls built.
Undrafted Gabela.
Encountered first Debug Log error "TryIssueJobPackage: NullReferenceException"
Unpaused
Hawk started documenting Beer Brewing, while the other pawns started constructing.
Paused
Drafted everyone
Unassigned Hawk's Study assignment via the Work tab
Undrafted everyone and unpaused
Encountered second Debug Log error "HumanResources.WorkGiver_DocumentTech provided target Thing_StudyDesk17423 but yielded no actual job for pawn Wolle. The CanGiveJob and JobOnX methods may not be synchronized".
All pawns still assigned to Study no longer constructing the wooden walls.
Pawns no longer assigned to Study resume constructing the wooden walls.

Thing_StudyDesk17423 is the study desk with the reserved Document Tech bill.

**Comment by jptrrs** [comment]
Thanks for reporting, archielecroy. 
I followed the steps you provided using the latest build, v0.29, but I was only able to reproduce the first error, which seems to happen only once when you unassign to study a pawn who's in the middle of a document task. That's both a completely harmless error and a very unlikely situation. But I could not get the second error and observed no issues with pawns not doing the expected job, regardless of the priority they're assigned to. 
If assigned to study as a priority and also to other worktypes, pawns should either go write books if there's an available desk for them to do so, or go do other assigned tasks according to their assigned work types.
At this point, the only reason I can think of why you're getting an error and I don't is a difference between game and mod versions.  The log doesn't tell me what's the RimWorld version you're using, but mine is 1.2. Regarding the mod, I didn't touch the workgivers between the one you had and the one I'm using, but there's always a possibility something else was reflected in that processs. Therefore, the only thing I can do is ask if you still get the same issue on the latest version. Do you?

**Comment by archielecroy** [comment]
Rimworld 1.2
Human Resources 0.29
Unable to reproduce

**Comment by jptrrs** [comment]
Ok, So I presume it didn`t happen again. That's one issue less.

**Comment by Reiquard** [comment]
Steps to reproduce:
 - Newly started scenario _(any)_
 - Built **[Study Desk]** with active bill **[Document technology]**
 - Built **[Book Shelf]**
 - **"Study"** set to highest priority for all pawns
 - **Pawn1** assigned for documentation **[Tech1]** and ordered to perform this task _(manually or via redrafting)_
 - Draft **Pawn1** after an **unfinished book** appears on the **[Study Desk]**
 - Assign **Pawn2** for documentation **[Tech2]** and redraft it => and here's the error:

``HumanResources.WorkGiver_DocumentTech provided target Thing_StudyDesk24174 but yielded no actual job for pawn Sachiko. The CanGiveJob and JobOnX methods may not be synchronized.
Verse.Log:Error(String, Boolean)
Verse.Log:ErrorOnce(String, Int32, Boolean)
RimWorld.JobGiver_Work:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.ThinkNode_PrioritySorter:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.ThinkNode_Priority:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.ThinkNode_Tagger:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.ThinkNode_Subtree:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.ThinkNode_Priority:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.ThinkNode_Conditional:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.ThinkNode_Priority:TryIssueJobPackage(Pawn, JobIssueParams)
Verse.AI.Pawn_JobTracker:DetermineNextJob(ThinkTreeDef&)
Verse.AI.Pawn_JobTracker:TryFindAndStartJob()
Verse.AI.Pawn_JobTracker:EndCurrentJob(JobCondition, Boolean, Boolean)
RimWorld.Pawn_DraftController:set_Drafted(Boolean)
RimWorld.Pawn_DraftController:<GetGizmos>b__13_1()
Verse.Command_Toggle:ProcessInput(Event)
Verse.GizmoGridDrawer:DrawGizmoGrid(IEnumerable`1, Single, Gizmo&)
RimWorld.InspectGizmoGrid:DrawInspectGizmoGridFor(IEnumerable`1, Gizmo&)
RimWorld.MainTabWindow_Inspect:DrawInspectGizmos()
RimWorld.InspectPaneUtility:ExtraOnGUI(IInspectPane)
RimWorld.MainTabWindow_Inspect:ExtraOnGUI()
Verse.WindowStack:WindowStackOnGUI()
RimWorld.UIRoot_Play:UIRootOnGUI()
Verse.Root:Verse.Root.OnGUI_Patch1(Root)
``

After that, **Pawn2** refuses to perform any task it has with the priority lower than documentation.

https://gist.github.com/e8866fbd1c5399277d074dcd6ffd91fc

**Comment by Reiquard** [comment]
After a long investigation, I'm pretty sure that `The CanGiveJob and JobOnX methods may not be synchronized.` refers to **HasJobOnThing** and **JobOnThing** methods and [changing](https://github.com/jptrrs/HumanResources/blob/f5b5c08585935b18d1ceefe9340e3e0ea9452a6f/Source/Work/WorkGiver_DocumentTech.cs#L36)
`return techComp.knownTechs.Where(x => !x.IsFinished).Intersect(techComp.homework).Any();`
to
`return base.HasJobOnThing(pawn, t) && techComp.knownTechs.Where(x => !x.IsFinished).Intersect(techComp.homework).Any();`
solves the problem with this error.

At least now **Pawn2** is not idle, having other priority tasks and behaves as if it were going to sew a hat on a tailoring bench, where there is already an unfinished hat of another pawn:
 - If there is an **unfinished hat(of Pawn1)**, while **Pawn1** **_is assigned_** as a Tailor via the work tab, **Pawn2** performs its other tasks.
 - If there is an **unfinished hat(of Pawn1)**, while **Pawn1** **_is not assigned_** as a Tailor via the work tab, **Pawn2** comes to the tailoring bench, hauls the **unfinished hat(of Pawn1)** nearby, and begins to sew the new hat.

The same with the **[Study Desk]** and the **books**, except that when **Pawn2** tries to perform the documentation job, the [error](https://pastebin.com/XukdRfUA) pops, but on the second attempt **Pawn2** documents the new book as expected.

Although, if I order the **Pawn1** to document its knowledge again (canceling the **Study** assignment for **Pawn2**), then after receiving _the above error_ again, it will go to the table, move the book aside, and start writing a **_completely new_** book of the same knowledge.
